src/api/fetcher.ts

```diff
@@ -0,0 +1,96 @@
+import { ENV } from './env';
+
+type FetcherOptions<T> = {
+  path: string;
+  query?: object;
+  body?: T;
+};
+
+type FetcherResponse<T> = Promise<T>;
+
+export const fetcher = {
+  get: async <T>({ path, query = {} }: FetcherOptions<T>): FetcherResponse<T> => {
+    return request<T>({
+      path,
+      query,
+      method: 'GET',
+    });
+  },
+  post: async <T>({ path, body }: FetcherOptions<T>): FetcherResponse<T> => {
+    return request<T>({
+      path,
+      body,
+      method: 'POST',
+      returnOriginalOnNoContent: true,
+    });
+  },
+  patch: async <T>({ path, body }: FetcherOptions<T>): FetcherResponse<T> => {
+    return request<T>({
+      path,
+      body,
+      method: 'PATCH',
+      returnOriginalOnNoContent: true,
+    });
+  },
+  delete: async <T>({ path }: FetcherOptions<T>): FetcherResponse<T> => {
+    return request<T>({
+      path,
+      method: 'DELETE',
+      returnOriginalOnNoContent: true,
+    });
+  },
+};
+
+type RequestOptions<T> = {
+  path: string;
+  method: string;
+  query?: object;
+  body?: T;
+  returnOriginalOnNoContent?: boolean;
+};
+
+const request = async <T>({
+  path,
+  method,
+  query,
+  body,
+  returnOriginalOnNoContent,
+}: RequestOptions<T>): Promise<T> => {
+  const url = new URL(ENV.BASE_URL + path);
+  Object.entries(query || {}).forEach(([key, value]) => {
+    if (value !== undefined && value !== null && String(value)) {
+      url.searchParams.append(key, String(value));
+    }
+  });
+
+  const headers = {
+    'Content-Type': 'application/json',
+    Authorization: `Basic ${ENV.TOKEN}`,
+  };
+
+  const config: RequestInit = {
+    method,
+    headers,
+  };
+
+  if (body && ['POST', 'PATCH'].includes(method)) {
+    config.body = JSON.stringify(body);
+  }
+
+  const response = await fetch(url, config);
+
+  if (!response.ok) {
+    // const errorData = await response.json().catch(() => null);
+    // throw new HttpError(response.status, errorData);
+    throw new Error();
+  }
+
+  if (response.status === 204 || response.headers.get('content-length') === '0') {
+    if (returnOriginalOnNoContent && body) {
+      return body as unknown as T;
+    }
+    return returnOriginalOnNoContent as unknown as T;
+  }
+
+  return await response.json();
```

Comment:
```suggestion
  return await response.json();
```

어차피 응답이 빈 값일꺼라서, await response.json() 으로 바로 넘어가도 차이가 없지 않을까요!?

Reply:
의견 주신 부분과 관련해서 궁금한 점이 있어 질문 드립니다!

바로 `await response.json()`을 호출하게 되면,
**응답 본문이 `{}` 같은 빈 객체가 아니라 아예 없을 경우**(`204 No Content`나 `content-length: 0`),
`SyntaxError: Unexpected end of JSON input` 같은 에러가 발생하진 않을까요?

Reply:
앗.. 그렇네요 ㅎㅎ
제가 잘못알고 있었네요 ㅠㅠ 죄송합니다.